Import publish profiles in the .NET Core SDK.

peterhuene opened the pull request:
This commit implements importing publish profiles in the .NET Core SDK.

Previously, there were two ways to get the publish profile respected when
publishing .NET Core applications:

* Using the Web SDK, which implicitly imports the `Microsoft.NET.Sdk.Publish`
SDK.
* Explicitly referencing the `Microsoft.NET.Sdk.Publish` SDK in the `Project
element` (e.g. `Project="Microsoft.NET.Sdk;Microsoft.NET.Sdk.Publish"`).

The problem is that the publish profile was being imported *after* many of the
important SDK properties (such as RuntimeIdentifier, SelfContained, and
UseAppHost) were defaulted by the .NET Core SDK.  Because the .NET Core SDK
does not expect the values of these properties to change once defaulted, users
are unable to set a particular property in the publish profile and get the
expected behavior in the .NET Core SDK.

For example, when setting just the `RuntimeIdentifier` in the publish profile,
it would publish without a RID-subdirectory, not as self-contained, and without
an apphost.  Contrast this to using the `--runtime` option for `dotnet
publish`, which uses a RID-subdirectory and publishes self-contained with an
apphost. Up until now, users had to work around this by setting *all* of the
desired properties in the publish profile that would otherwise be defaulted by
the .NET Core SDK.

The Web SDK fixed this by importing the publish profile *before* the .NET Core
SDK was imported (see aspnet/websdk#534).  However, this could not fix the
problem for users that have an existing project that directly references the
`Microsoft.NET.Sdk.Publish` SDK.

To fix that, the .NET Core SDK will not attempt to import the publish profile
if it hasn't already been imported by the Web SDK.  If the .NET Core SDK
successfully imports the publish profile, it sets a property that prevents the
Web SDK from attempting to import it.

Fixes #10647.

Commit:
Import publish profiles in the .NET Core SDK.

This commit implements importing publish profiles in the .NET Core SDK.

Previously, there were two ways to get the publish profile respected when
publishing .NET Core applications:

* Using the Web SDK, which implicitly imports the `Microsoft.NET.Sdk.Publish`
SDK.
* Explicitly referencing the `Microsoft.NET.Sdk.Publish` SDK in the `Project
element` (e.g. `Project="Microsoft.NET.Sdk;Microsoft.NET.Sdk.Publish"`).

The problem is that the publish profile was being imported *after* many of the
important SDK properties (such as RuntimeIdentifier, SelfContained, and
UseAppHost) were defaulted by the .NET Core SDK.  Because the .NET Core SDK
does not expect the values of these properties to change once defaulted, users
are unable to set a particular property in the publish profile and get the
expected behavior in the .NET Core SDK.

For example, when setting just the `RuntimeIdentifier` in the publish profile,
it would publish without a RID-subdirectory, not as self-contained, and without
an apphost.  Contrast this to using the `--runtime` option for `dotnet
publish`, which uses a RID-subdirectory and publishes self-contained with an
apphost. Up until now, users had to work around this by setting *all* of the
desired properties in the publish profile that would otherwise be defaulted by
the .NET Core SDK.

The Web SDK fixed this by importing the publish profile *before* the .NET Core
SDK was imported (see aspnet/websdk#534).  However, this could not fix the
problem for users that have an existing project that directly references the
`Microsoft.NET.Sdk.Publish` SDK.

To fix that, the .NET Core SDK will not attempt to import the publish profile
if it hasn't already been imported by the Web SDK.  If the .NET Core SDK
successfully imports the publish profile, it sets a property that prevents the
Web SDK from attempting to import it.

Fixes #10647.

peterhuene commented:
@vijayrkn apologies for the delay; here's the .NET Core side of the fix.

vijayrkn commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.ImportPublishProfile.targets:
```diff
@@ -0,0 +1,31 @@
+    <_PublishProfileRootFolder Condition="'$(_PublishProfileRootFolder)' == ''">$(MSBuildProjectDirectory)\$(_PublishProfileDesignerFolder)\PublishProfiles\</_PublishProfileRootFolder>
+    <PublishProfile Condition="'$(PublishProfile)' ==''">FileSystem</PublishProfile>
+    <PublishProfileName Condition="'$(PublishProfileName)' == ''">$([System.IO.Path]::GetFileNameWithoutExtension($(PublishProfile)))</PublishProfileName>
+    <PublishProfileFullPath Condition="'$(PublishProfileFullPath)' == ''">$(_PublishProfileRootFolder)$(PublishProfileName).pubxml</PublishProfileFullPath>
```
The check for Exists for PublishProfileFullPath below should handle the default import. So we are good with the default scenarios. https://github.com/aspnet/websdk/blob/ef0b2621e90ed62b30b72d337c317f61f3df0c12/src/Publish/Sdk/ImportPublishProfile.targets#L20

Commit:
Add test to cover Microsoft.NET.Sdk.Publish use case.

This commit adds a test that covers referencing the `Microsoft.NET.Sdk.Publish`
SDK directly and having the publish profile property imported.

vijayrkn commented:
Thanks @peterhuene for getting this change. When you have a build with this fix, can you please share so that we can test the VS and commandline scenarios ?

peterhuene commented:
Will do!

peterhuene commented:
Looks like a whole slew of test failures due to `error NU1102: Unable to find package Microsoft.AspNetCore.App.Runtime.win-x64 with version (= 3.0.0-preview6-19226-02)`.

vijayrkn commented:
> Looks like a whole slew of test failures due to error NU1102: Unable to find package Microsoft.AspNetCore.App.Runtime.win-x64 with version (= 3.0.0-preview6-19226-02).

The above change shouldn't have caused this.

peterhuene commented:
Agreed.  Possibly something is up with the feeds.  I'll look into it Monday.

wli3 commented on src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishAHelloWorldProject.cs:
```diff
@@ -539,5 +539,89 @@ public void It_contains_no_duplicates_in_resolved_publish_assets()
+            var output = command.GetOutputDirectory(targetFramework: tfm, runtimeIdentifier: rid);
+
+            output.Should().HaveFiles(new[] {
+                $"{testProject.Name}.dll",
```
This looks like a generic output assertion (with publishing profile or without should be the same?). What really matters here? I guess as long as it succeeds?

peterhuene replied:
These are files that the publish profile doesn't affect, but it ensures the publish succeeded to the right (RID-based) path.

wli3 commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.ImportPublishProfile.targets:
```diff
@@ -0,0 +1,31 @@
+<Project ToolsVersion="14.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <PropertyGroup>
+    <!-- Default to having imported the publish profile so the Web SDK doesn't also attempt to do so. -->
+    <PublishProfileImported>true</PublishProfileImported>
```
There is a lot of default property. If any one of these property is wrong, it will fail the build? If so the test is good enough

peterhuene replied:
Altering most of these properties would result in the wrong path to the import profile the tests that were added create, so it will cause them to fail because the wrong output from the publish will occur (primarily the publish directory won't exist since it'll be RID-less).

peterhuene commented:
@wli3 are there any additional comments you have on this PR or follow-ups to the above questions?

wli3 commented:
Sorry i think i missed your reply notification
